src/tools/immutable_object_cache/Types.h

```diff
@@ -0,0 +1,60 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+
+#ifndef CEPH_CACHE_TYPES_H
+#define CEPH_CACHE_TYPES_H
+
+#include "include/encoding.h"
+#include "include/Context.h"
+
+namespace ceph {
+namespace immutable_obj_cache {
+
+struct ObjectCacheMsgHeader {
+    uint64_t seq;                         /* sequence id */
+    uint16_t type;                        /* msg type */
+    uint16_t version;                     /* object cache version */
+    uint32_t padding;
+    uint32_t data_len;
+    uint32_t reserved;
+
+    void encode(bufferlist& bl) const;
+    void decode(bufferlist::const_iterator& it);
+};
+
+class ObjectCacheMsgData {
+public:
+  uint64_t m_image_size;
+  uint64_t m_read_offset;
+  uint64_t m_read_len;
+  std::string m_pool_name;
+  std::string m_image_name;
+  std::string m_oid;
+
+  void encode(bufferlist& bl);
+  void decode(bufferlist& bl);
+};
+
+class ObjectCacheRequest {
```

Comment:
I would like to see a messaging structure that doesn't attempt to handle all cases with a single generic struct. At a high-level. you would just use ```ENCODE_START```/```ENCODE_FINISH``` and ```DECODE_START```/```DECODE_FINISH``` pairs to encode a version, backwards compatible version, and overall message length into the bufferlist. Then you can encode/decode a message-type enum to permit very specific messages. See [1] for an example pattern that we use all over RBD.

[1] https://github.com/ceph/ceph/blob/master/src/librbd/mirroring_watcher/Types.cc#L81

Reply:
@dillaman thanks a lot for the example! In the current implementation each messag has a header [ObjectCacheMsgHeader](https://github.com/ceph/ceph/blob/3bee7bd9ea08de18e213a69a98d639955387d4a0/src/tools/immutable_object_cache/Types.cc#L13-L33) and the body [ObjectCacheMsgData](https://github.com/ceph/ceph/blob/3bee7bd9ea08de18e213a69a98d639955387d4a0/src/tools/immutable_object_cache/Types.cc#L35-L58). The header contains some heavily used info like sequence id, message type, body length. The body has the info of the target object to look up in cache. The reply message body will include the cache file path so the hook can read data from if it's a cache hit. I added the version/compat version macro in the encode/decode func.  I'm not sure if I get right usage, could you please help to take a look if you have time? here's the commit https://github.com/ceph/ceph/pull/25545/commits/3bee7bd9ea08de18e213a69a98d639955387d4a0

Reply:
pushed a new commit to clean up https://github.com/ceph/ceph/pull/25545/commits/d4d3dc76e4a457144dd7d4f541557688c3f32960